lib/router.js

```diff
@@ -152,6 +154,40 @@ class Router {
         }
     }
 
+    _getCachedModule(modDef, globals) {
```

Comment:
This function doesn't really 'get', it also updates. very not intuitive. Something like getCachedOrSet would be nicer.

Reply:
All it does it gets a module. `this._modules` is now a map of maps: the first layer has the module definitions digests as keys with values being maps of pairs (globals digest, actual module). So, what this method does is it checks that the higher-lavel mapping exists, and if it doesn't it creates it because we know we will store the actual module there afterwards.

Reply:
1. Why do we need a map of maps? Wouldn't a single-level map be easier? just append module signature and globals signature and we'd have the same result wouldn't we?
2. Usually 'get' operations don't have side effects.
3. I think it would be more natural to have `getModuleDigest` function and then just use a single-level Map for caching with the digest generated by that function.

Reply:
> 1. Why do we need a map of maps? Wouldn't a single-level map be easier? just append module signature and globals signature and we'd have the same result wouldn't we?

I find it a better and less-confusing approach. Sure, we could use one map and keys like `${modDefDigest}/${globalsDigest}` but I don't see a compelling reason to do so. Having a hierarchy implies the natural way of thinking about this: we group modules first by their definitions and then by the context (aka globals). Memory is not really a concern here (it's not much more expensive to have extra maps) since after the spec loading is complete we wipe out the cache.

> 2. Usually 'get' operations don't have side effects.

If you are referring to adding a new map in case there is not one yet, I wouldn't call that a _side effect_; we are simply spearing ourselves to do a check and an addition further down the line, because we know that we will add that map no matter what. There are only two outcomes: the module is already cached or it isn't. In the former case, we return it, but in the latter we cache it after `require()`-ing it.

> 3. I think it would be more natural to have `getModuleDigest` function and then just use a single-level Map for caching with the digest generated by that function.

See my reasoning under (1) above. I started out with a single-level map but didn't really like it. Having a hierarchy also allows us to have it clear that the module we are caching depends on both the definition and the context (in a previous version of the code I actually had a bug where I was using only the module definition to cache the module and later realised that's no bueno).